README.md
# python-logging-rabbitmq

![Package License](https://img.shields.io/github/license/albertomr86/python-logging-rabbitmq)
![PyPI - Version](https://img.shields.io/pypi/v/python_logging_rabbitmq)
![PyPI - Wheel](https://img.shields.io/pypi/wheel/python_logging_rabbitmq)
![Build Status](https://github.com/albertomr86/python-logging-rabbitmq/workflows/ci/badge.svg)


Logging handler to ships logs to RabbitMQ. Compatible with Django.

## Installation
Install using pip.

```sh
pip install python_logging_rabbitmq
```

### Versions

| Version      | Dependency      |
|--------------|-----------------|
| &gt;= 2.x    |  Pika == 0.13   |
| <= 1.1.1     |  Pika <= 0.10   |


## Handlers
This package has two built-in handlers that you can import as follows:

```python
from python_logging_rabbitmq import RabbitMQHandler
```

or (thanks to [@wallezhang](https://github.com/wallezhang))

```python
from python_logging_rabbitmq import RabbitMQHandlerOneWay
```

| Handler               | Description                                                                                                                                                                                                                                                                                  |
|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RabbitMQHandler       | Basic handler for sending logs to RabbitMQ. Every record will be delivered directly to RabbitMQ using the exchange configured.                                                                                                                                                               |
| RabbitMQHandlerOneWay | High throughput handler. Initializes an internal queue where logs are stored temporarily. A thread is used to deliver the logs to RabbitMQ using the exchange configured. Your app doesn't need to wait until the log is delivered. Notice that if the main thread dies you might lose logs. |

## Standalone python
To use with python first create a logger for your app, then create an instance of the handler and add it to the logger created.

```python
import logging
from python_logging_rabbitmq import RabbitMQHandler

logger = logging.getLogger('myapp')
logger.setLevel(logging.DEBUG)

rabbit = RabbitMQHandler(host='localhost', port=5672)
logger.addHandler(rabbit)

logger.debug('test debug')
```

As result, a similar message as follows will be sent to RabbitMQ:

```json
{
	"relativeCreated":280.61580657958984,
	"process":13105,
	"args":[],
	"module":"test",
	"funcName":"<module>",
	"host":"albertomr86-laptop",
	"exc_text":null,
	"name":"myapp",
	"thread":140032818181888,
	"created":1482290387.454017,
	"threadName":"MainThread",
	"msecs":454.01692390441895,
	"filename":"test.py",
	"levelno":10,
	"processName":"MainProcess",
	"pathname":"test.py",
	"lineno":11,
	"msg":"test debug",
	"exc_info":null,
	"levelname":"DEBUG"
}
```

## Sending logs
By default, logs will be sent to RabbitMQ using the exchange **'log'**, this should be of **type topic**. The **routing key** used is formed by concatenating the *logger name* and the *log level*. For example:

```python
import logging
from python_logging_rabbitmq import RabbitMQHandler

logger = logging.getLogger('myapp')
logger.setLevel(logging.DEBUG)
logger.addHandler(RabbitMQHandler(host='localhost', port=5672))

logger.info('test info')
logger.debug('test debug')
logger.warning('test warning')
```

The messages will be sent using the following routing keys:
- myapp.INFO
- myapp.DEBUG
- myapp.WARNING

For an explanation about topics and routing keys go to https://www.rabbitmq.com/tutorials/tutorial-five-python.html

When create the handler, you're able to specify different parameters in order to connect to RabbitMQ or configure the handler behavior.

## Overriding routing-key creation

If you wish to override routing-key format entirely, you can pass
`routing_key_formatter` function which takes `LogRecord` objects and returns
routing-key. For example:

```python
RabbitMQHandler(
	host='localhost',
	port=5672,
	routing_key_formatter=lambda r: (
		'some_exchange_prefix.{}'.format(r.levelname.lower())
	)
)
```

## Configuration
These are the configuration allowed:

| Parameter             | Description                                                                                                                              | Default                               |
|-----------------------|------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------|
| host                  | RabbitMQ Server hostname or ip address.                                                                                                  | localhost                             |
| port                  | RabbitMQ Server port.                                                                                                                    | 5672                                  |
| username              | Username for authentication.                                                                                                             | None                                  |
| password              | Provide a password for the username.                                                                                                     | None                                  |
| exchange              | Name of the exchange to publish the logs. This exchange is considered of type topic.                                                     | log                                   |
| declare_exchange      | Whether or not to declare the exchange.                                                                                                  | False                                 |
| remove_request        | If True (default), remove request & exc info.                                                                                            | True                                  |
| routing_key_format    | Customize how messages are routed to the queues.                                                                                         | {name}.{level}                        |
| routing_key_formatter | Customize how routing-key is constructed.                                                                                                | None                                  |
| connection_params     | Allow extra params to connect with RabbitMQ.                                                                                             | None                                  |
| formatter             | Use custom formatter for the logs.                                                                                                       | python_logging_rabbitmq.JSONFormatter |
| close_after_emit      | Close the active connection after send a log. A new connection is open for the next log.                                                 | False                                 |
| fields                | Dict to add as a field in each logs send to RabbitMQ. This is useful when you want fields in each log but without pass them every time.  | None                                  |
| fields_under_root     | When is True, each key in parameter 'fields' will be added as an entry in the log, otherwise they will be logged under the key 'fields'. | True                                  |
| message_headers       | A dictionary of headers to be published with the message.                                                                                | None                                  |
| record_fields         | A set of attributes that should be preserved from the record object.                                                                     | None                                  |
| exclude_record_fields | A set of attributes that should be ignored from the record object.                                                                       | None                                  |
| heartbeat             | Lower bound for heartbeat timeout.                                                                                                       | 60                                    |
| content_type          | The format of the message sent to the queue.                                                                                             | text/plain                            |


### Examples
#### RabbitMQ Connection

```python
rabbit = RabbitMQHandler(
	host='localhost',
	port=5672,
	username='guest',
	password='guest',
	connection_params={
		'virtual_host': '/',
		'connection_attempts': 3,
		'socket_timeout': 5000
	}
)
```

#### Custom fields

```python
rabbit = RabbitMQHandler(
	host='localhost',
	port=5672,
	fields={
		'source': 'MyApp',
		'env': 'production'
	},
	fields_under_root=True
)
```

#### Custom formatter
By default, python_logging_rabbitmq implements a custom JSONFormatter; but if you prefer to format your own message you could do it as follow:

```python
import logging
from python_logging_rabbitmq import RabbitMQHandler

FORMAT = '%(asctime)-15s %(message)s'
formatter = logging.Formatter(fmt=FORMAT)
rabbit = RabbitMQHandler(formatter=formatter)
```

For a custom JSON Formatter take a look at https://github.com/madzak/python-json-logger

## Django
To use with Django add the handler in the [logging config](https://docs.djangoproject.com/en/1.9/topics/logging/#configuring-logging).

```python
LOGGING = {
	'version': 1,
	'disable_existing_loggers': False,
	'handlers': {
		'rabbit': {
			'level': 'DEBUG',
			'class': 'python_logging_rabbitmq.RabbitMQHandler',
			'host': 'localhost'
		}
	},
	'loggers': {
		'myapp': {
			'handlers': ['rabbit'],
			'level': 'DEBUG',
			'propagate': False
		}
	}
}
```

## Configuration
Same as when use it with standalone python, you could configure the handle directly when declaring it in the config:

```python
LOGGING = {
	'version': 1,
	'disable_existing_loggers': False,
	'handlers': {
		'rabbit': {
			'level': 'DEBUG',
			'class': 'python_logging_rabbitmq.RabbitMQHandler',
			'host': 'localhost',
			'port': 5672,
			'username': 'guest',
			'password': 'guest',
			'exchange': 'log',
			'declare_exchange': False,
			'connection_params': {
				'virtual_host': '/',
				'connection_attempts': 3,
				'socket_timeout': 5000
			},
			'fields': {
				'source': 'MainAPI',
				'env': 'production'
			},
			'fields_under_root': True
		}
	},
	'loggers': {
		'myapp': {
			'handlers': ['rabbit'],
			'level': 'DEBUG',
			'propagate': False
		}
	}
}
```

#### Custom formatter

```python
LOGGING = {
	'version': 1,
	'disable_existing_loggers': False,
	'formatters': {
		'standard': {
			'format': '%(levelname)-8s [%(asctime)s]: %(message)s'
		}
	},
	'handlers': {
		'rabbit': {
			'level': 'DEBUG',
			'class': 'python_logging_rabbitmq.RabbitMQHandler',
			'host': 'localhost',
			'formatter': 'standard'
		}
	},
	'loggers': {
		'myapp': {
			'handlers': ['rabbit'],
			'level': 'DEBUG',
			'propagate': False
		}
	}
}
```

#### JSON formatter

```sh
pip install python-json-logger
```

```python
LOGGING = {
	'version': 1,
	'disable_existing_loggers': False,
	'formatters': {
		'json': {
			'()': 'pythonjsonlogger.jsonlogger.JsonFormatter',
			'fmt': '%(name)s %(levelname) %(asctime)s %(message)s'
		}
	},
	'handlers': {
		'rabbit': {
			'level': 'DEBUG',
			'class': 'python_logging_rabbitmq.RabbitMQHandler',
			'host': 'localhost',
			'formatter': 'json'
		}
	},
	'loggers': {
		'myapp': {
			'handlers': ['rabbit'],
			'level': 'DEBUG',
			'propagate': False
		}
	}
}
```

## Releases

[CHANGELOG](./CHANGELOG.md)

## Similar efforts
- MQHandler (https://github.com/ziXiong/MQHandler)
- http://stackoverflow.com/a/25479008/2716524
- LogBook (http://pydoc.net/Python/Logbook/0.12.5/logbook.queues/)
- https://pypi.python.org/pypi/py-amqp-logging/0.1


CHANGELOG.md
# v2.3.0 (2023-09-28)

### Features

- PR#24 : Added the option `remove_request` to decide whether the request's properties are removed from the record to log. (Thank you [@donbowman](https://github.com/donbowman))

### Fixes

- PR#37 : Call `task_done()` to complete the item dequeued. Reported by PR#32 PR#31, and PR#30.


# v2.2.0 (2022-04-24)

### Features

- PR#25 Handling thread shutdown (Thanks to [@donbowman](https://github.com/donbowman)).


# v2.1.0 (2022-04-23)

### Features

- PR#19 Heartbeat.
- PR#26 Put messages in the queue with a specific content type.

### Fixes

- PR#27 Minor corrections in the documentation.


# v2.0.0 (2019-06-26)

### Fixes

- PR#14 Heartbeat interval

# v1.1.1 (2019-03-10)

### Features

- Integration with [Travis CI](https://travis-ci.org/).
- Configuration for tests. Using [pipenv](https://github.com/pypa/pipenv).
- PR#10 Support for `pika` > 0.10.

### Fixes

- Removed direct dependency with Django.
- PR#11 The log record was being formatted twice.


# v1.0.9 (2018-05-04)

### Fixes

- PR#7 Fixed `exchange_type` parameter in `channel.exchange_declare` (Thanks to [@cklos](https://github.com/cklos)).


# v1.0.8 (2018-03-22)

### Features

- PR#6 Allow message headers when publishing to the queue (Thanks to [@merretbuurman](https://github.com/merretbuurman)).


# v1.0.7 (2017-05-15)

### Features

- Added support to customize the `routing_key` (Thanks to [@hansyulian](https://github.com/hansyulian)).


# v1.0.6 (2017-03-29)

### Fixes

- PR#3, PR#4 Fix compatibility with python3 in `RabbitMQHandlerOneWay` (by [@sactre](https://github.com/sactre)).


# v1.0.5 (2017-03-28)

### Chrores

- Explicit local imports.


# v1.0.4 (2017-03-15)

### Features

- PR#1 Added new handler `RabbitMQHandlerOneWay` (by [@wallezhang](https://github.com/wallezhang)). 


# v1.0.3 (2017-03-13)

- Added config parameter `close_after_emit`. Close connection after emit the record.


# v1.0.2 (2016-12-23)

### Features

- Expose `JSONFormatter`.


# v1.0.1 (2016-12-21)

Minor corrections in the documentation.


# v1.0.0 (2016-12-21)

First release.



python_logging_rabbitmq/handlers_oneway.py
# -*- coding: utf-8 -*-
import time
import logging
import threading
from copy import copy

import pika
from pika import credentials

from .compat import Queue, QueueEmpty
from .filters import FieldFilter
from .formatters import JSONFormatter
from .compat import ExceptionReporter


class RabbitMQHandlerOneWay(logging.Handler):
    """
    Python/Django logging handler to ship logs to RabbitMQ.
    Inspired by: https://github.com/ziXiong/MQHandler
    """
    def __init__(self, level=logging.NOTSET, formatter=None,
        host='localhost', port=5672, connection_params=None,
        username=None, password=None,
        exchange='log', declare_exchange=False,
        remove_request=True,
        routing_key_format="{name}.{level}",
        routing_key_formatter=None,
        close_after_emit=False,
        fields=None, fields_under_root=True, message_headers=None,
        record_fields=None, exclude_record_fields=None,
        heartbeat=60,
        content_type='text/plain'):
        # Initialize the handler.
        #
        # :param level:                 Logs level.
        # :param formatter:             Use custom formatter for the logs.
        # :param host:                  RabbitMQ host. Default localhost
        # :param port:                  RabbitMQ Port. Default 5672
        # :param connection_params:     Allow extra params to connect with RabbitMQ.
        # :param message_headers:       A dictionary of headers to be published with the message. Optional.
        # :param username:              Username in case of authentication.
        # :param password:              Password for the username.
        # :param exchange:              Send logs using this exchange.
        # :param declare_exchange:      Whether or not to declare the exchange.
        # :param remove_request:        If true (default), remove request/exc info
        # :param routing_key_format:    Customize how messages will be routed to the queues.
        # :param routing_key_formatter: Override how messages will be routed to the queues.
        #                               Formatter is passed record object.
        # :param close_after_emit:      Close connection after emit the record?
        # :param fields:                Send these fields as part of all logs.
        # :param fields_under_root:     Merge the fields in the root object.
        # :record_fields:               A set of attributes that should be preserved from the record object.
        # :exclude_record_fields:       A set of attributes that should be ignored from the record object.
        # :heartbeat:                   Lower bound for heartbeat timeout.
        # :content_type:                The format of the message sent to the queue.

        super(RabbitMQHandlerOneWay, self).__init__(level=level)

        # Important instances/properties.
        self.exchange = exchange
        self.connection = None
        self.channel = None
        self.exchange_declared = not declare_exchange
        self.remove_request = remove_request
        self.routing_key_format = routing_key_format
        self.close_after_emit = close_after_emit

        # Connection parameters.
        # Allow extra params when connect to RabbitMQ.
        # @see: http://pika.readthedocs.io/en/0.10.0/modules/parameters.html#pika.connection.ConnectionParameters
        conn_params = connection_params if isinstance(connection_params, dict) else {}
        self.connection_params = conn_params.copy()
        self.connection_params.update(dict(host=host, port=port, heartbeat=heartbeat, blocked_connection_timeout=150))

        if username and password:
            self.connection_params['credentials'] = credentials.PlainCredentials(username, password)

        # Extra params for message publication
        self.message_headers = message_headers
        self.content_type = content_type

        # Save routing-key formatter.
        self.routing_key_formatter = routing_key_formatter

        # Logging.
        self.formatter = formatter or JSONFormatter(
            include=record_fields,
            exclude=exclude_record_fields
        )
        self.fields = fields if isinstance(fields, dict) else {}
        self.fields_under_root = fields_under_root

        if len(self.fields) > 0:
            self.addFilter(FieldFilter(self.fields, self.fields_under_root))

        # Connect.
        self.createLock()

        # message queue
        self.queue = Queue()
        self.start_message_worker()

    def open_connection(self):
        """
        Connect to RabbitMQ.
        """
        # Set logger for pika.
        # See if something went wrong connecting to RabbitMQ.
        if not self.connection or self.connection.is_closed or not self.channel or self.channel.is_closed:
            handler = logging.StreamHandler()
            handler.setFormatter(self.formatter)
            rabbitmq_logger = logging.getLogger('pika')
            rabbitmq_logger_propagate = rabbitmq_logger.propagate
            rabbitmq_logger_level = rabbitmq_logger.level

            rabbitmq_logger.addHandler(handler)
            rabbitmq_logger.propagate = False
            rabbitmq_logger.setLevel(logging.WARNING)

            # Connect.
            if not self.connection or self.connection.is_closed:
                self.connection = pika.BlockingConnection(pika.ConnectionParameters(**self.connection_params))

            if not self.channel or self.channel.is_closed:
                self.channel = self.connection.channel()

            if self.exchange_declared is False:
                self.channel.exchange_declare(exchange=self.exchange, exchange_type='topic', durable=True, auto_delete=False)
                self.exchange_declared = True

            # Manually remove logger to avoid shutdown message.
            rabbitmq_logger.propagate = rabbitmq_logger_propagate
            rabbitmq_logger.setLevel(rabbitmq_logger_level)
            rabbitmq_logger.removeHandler(handler)

    def close_connection(self):
        """
        Close active connection.
        """
        self.stopping.set()
        while not self.stopped.is_set():
            time.sleep(1)
        if self.channel:
            del self.channel

        if self.connection:
            del self.connection

        self.connection, self.channel = None, None

    def start_message_worker(self):
        self.stopping = threading.Event()
        self.stopped = threading.Event()
        worker = threading.Thread(target=self.message_worker)
        worker.setDaemon(True)
        worker.start()

    def message_worker(self):
        while not self.stopping.is_set():
            try:
                record, routing_key = self.queue.get(block=True, timeout=10)
                
                try:
                    if not self.connection or self.connection.is_closed or not self.channel or self.channel.is_closed:
                        self.open_connection()

                    self.channel.basic_publish(
                        exchange=self.exchange,
                        routing_key=routing_key,
                        body=record,
                        properties=pika.BasicProperties(
                            delivery_mode=2,
                            headers=self.message_headers,
                            content_type=self.content_type
                        )
                    )
                finally:
                    if hasattr(self, 'queue'):
                        self.queue.task_done()

            except QueueEmpty:
                continue
            except Exception:
                self.channel, self.connection = None, None
                self.handleError(record)
            finally:
                if self.stopping.is_set():
                    self.stopped.set()
                    break
                if self.close_after_emit:
                    self.close_connection()
        self.stopped.set()

    def emit(self, record):
        if not hasattr(self, 'queue') or self.stopped.is_set():
            return

        try:
            if self.routing_key_formatter:
                routing_key = self.routing_key_formatter(record)
            else:
                routing_key = self.routing_key_format.format(
                    name=record.name,
                    level=record.levelname
                )

            if self.remove_request and hasattr(record, 'request'):
                no_exc_record = copy(record)
                del no_exc_record.exc_info
                del no_exc_record.exc_text
                del no_exc_record.request

                if record.exc_info:
                    exc_info = record.exc_info
                else:
                    exc_info = (None, record.getMessage(), None)

                if ExceptionReporter:
                    reporter = ExceptionReporter(record.request, is_email=False, *exc_info)
                    no_exc_record.traceback = reporter.get_traceback_text()

                formatted = self.format(no_exc_record)
            else:
                formatted = self.format(record)

            self.queue.put((formatted, routing_key))
        except Exception:
            self.channel, self.connection = None, None
            self.handleError(record)

    def close(self):
        """
        Free resources.
        """
        self.acquire()

        if hasattr(self, 'queue'):
            del self.queue

        try:
            self.close_connection()
        finally:
            self.release()
